Different way of setting up modular spacing

yvonnetangsu opened the pull request:
# READY FOR REVIEW

# Summary
- Use the nested sass map approach for modular spacing allowing for easier finetuning and updating of individual values.
- New sass map `$su-responsive-spacing-map` in `variables/core/_responsive-spacing.scss`
- Add new mixin `@responsive-spacing` instead of replacing the old one per Decanter meeting discussion.
- Consider removing the negative modular spacing steps, since the values are 1px apart between adjacent device sizes. Maybe no need for modular spacing at those small ends (just pick the middle value for all devices)?
- This PR also includes some helper functions, e.g., a "map-deep-get" function that CSS Tricks has featured to retrieve values from nested sass maps.

# Needed By (Date)
- When does this need to be merged by?

# Urgency
- How critical is this PR?

# Steps to Test

1. Pull this branch
1. Apply `@include responsive-spacing('padding', 4 3 null 10);` for example to a component
1. Apply `@include responsive-spacing('margin', 8);` for example to a component
1. Verify that it's pulling from the new responsive spacing value in the below sass map (it should render 3 set of spacing values for the 3 breakpoint ranges - xs-sm, md-xl (768px+), 2xl (1500px+))
![Screen Shot 2020-10-08 at 5 14 04 PM](https://user-images.githubusercontent.com/42749717/95527260-b4989480-0989-11eb-8ffa-f0309dbfb18e.png)


# Affected Projects or Products
- Does this PR impact any particular projects, products, or modules?

# Associated Issues and/or People
- #724

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

sherakama commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr704-artqrbwckdrvxjsngf6uaw9xrixwdmzq.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5f78290f8ff35823c4ef06f8

sherakama commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr704-artqrbwckdrvxjsngf6uaw9xrixwdmzq.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5f78290f8ff35823c4ef06f8

sherakama commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr704-artqrbwckdrvxjsngf6uaw9xrixwdmzq.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5f78290f8ff35823c4ef06f8

sherakama commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr704-artqrbwckdrvxjsngf6uaw9xrixwdmzq.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5f78290f8ff35823c4ef06f8

yvonnetangsu commented:
@katrialesser @sherakama This is ready for the two of you to take a look when you have time. I'm not replacing the old @modular-spacing since it's used in so many places, instead, just created @modular-spacing-new which pulls from the new values. Let me know what you think.

sherakama commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr704-artqrbwckdrvxjsngf6uaw9xrixwdmzq.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5f78290f8ff35823c4ef06f8

sherakama commented on core/src/scss/utilities/functions/_map-deep-get.scss:
```diff
@@ -0,0 +1,12 @@
+/// @param {Map} $map - Map
+/// @param {Arglist} $keys - Key chain
+/// @return {*} - Desired value
+@function map-deep-get($map, $keys...) {
```
Cool. I like this.

yvonnetangsu replied:
Thanks to Hugo Giraudel 😄

sherakama commented on core/src/scss/utilities/mixins/breakpoint/_grid-media-device.scss:
```diff
@@ -0,0 +1,20 @@
+/// @param {key code} $breakpoint - The breakpoint shortcode - 'mobile', 'tablet' or 'desktop'
+///
+/// @group mixin
+@mixin grid-media-device($breakpoint: 'tablet') {
```
I can see this being a handy thing to use.

sherakama commented on core/src/scss/utilities/variables/core/_modular-spacing.scss:
```diff
@@ -46,3 +46,27 @@ $su-modular-spacing-multiplier: (
+///
+/// @group variable
+$su-modular-spacing-map: (
+  -4: (mobile: 0.3rem, tablet: 0.3rem, desktop: 0.3rem), // Consider removing
+  -3: (mobile: 0.5rem, tablet: 0.6rem, desktop: 0.6rem), // Consider removing
+  -2: (mobile: 0.8rem, tablet: 0.9rem, desktop: 1rem), // Consider removing
+  -1: (mobile: 1.1rem, tablet: 1.2rem, desktop: 1.3rem), // Consider removing
```
These could be useful on large things.

yvonnetangsu replied:
If it's `em`, then I'd agree, but it's `rem` so it's only 1 pixel difference (or no difference) in most of these above 👀

sherakama commented on core/src/scss/utilities/mixins/modular-scale/_modular-spacing-new.scss:
```diff
@@ -0,0 +1,77 @@
+/// @param {map} $grid - Breakpoint device sass map. Defaults to $su-grid-device.
+///
+/// @group mixin
+@mixin modular-spacing-new($type: 'padding', $steps: 0, $grid: $su-grid-device) {
```
I'd much rather not have a `-new` function but instead, change the original. We do want to change the spacing of everything (if even just a little). We can make this a minor release and have release notes about how to upgrade. If the change in the scale is too drastic and we cannot change this function to use the new values then we must deprecate it and re-name this function something else as `-new` gets old quick. Perhaps just `modular-space`

yvonnetangsu replied:
I can see how the `-new` can get old real quick 😂   I'm just a little afraid of replacing it right away, since it's used everywhere including our masthead, global footer etc.

yvonnetangsu replied:
Btw, I believe @kerri-augenstein said these values still need to be finetuned. I supposed it would be interesting to just replace the mixin in the branch and see what break and what work well, and tweak accordingly 🤔

sherakama replied:
Double-checking with @kerri-augenstein here then. We DO want to change the modular spacing values everywhere we used them to the new ones? I guess my question is, how much different is say, a 4, from the previous function to the new one? If the adjustment is too big or a 4 yesterday should really be 5 tomorrow we need to talk about the DX of this change. We can move SOME cheese in this version, but we might have to run with the new function and deprecate the existing.

yvonnetangsu replied:
Oh, some of the changes are pretty drastic (more so for the larger steps). But yeah, I anticipate some back and forth with @kerri-augenstein here.

yvonnetangsu replied:
@sherakama I pushed up a branch `modular-spacing-map-replace`, which straight up replace the old mixin with this new one. I'm testing it on WP Redwood....strangely, working surprisingly well. Not really breaking things except in a few expected places (that's specific to that theme). Need more testing of course. Feel free to pull that branch and test it out on D8.

sherakama commented on core/src/scss/utilities/mixins/modular-scale/_modular-spacing-new.scss:
```diff
@@ -0,0 +1,77 @@
+///
+/// @group mixin
+@mixin modular-spacing-new($type: 'padding', $steps: 0, $grid: $su-grid-device) {
+  @each $key, $breakpoint in $grid {
```
Why the change to only the three steps? We do some fine tweaking at various breakpoints for spacing on top of the modular spacing rules for some of our custom stuff. Wouldn't we still want the granularity or is that just causing too many media queries and specificity?

yvonnetangsu replied:
Oh, this part is actually unchanged - I believe the current modular-spacing mixin only writes 3 sets of values (since there are only 3 unique values in modular-spacing-base).

sherakama replied:
AFAICT the current modular spacing uses the `$su-grid-media` as the breakpoint loop. It should have a value at each breakpoint.

```
$su-grid-media: (
  'xs': $su-breakpoint-xs,
  'sm': $su-breakpoint-sm,
  'md': $su-breakpoint-md,
  'lg': $su-breakpoint-lg,
  'xl': $su-breakpoint-xl,
  '2xl': $su-breakpoint-2xl
) !default;
```

yvonnetangsu replied:
Heh I can answer that one. I added a comparison in the current mixin to compare the spacing value computed for that breakpoint. If it's the same value as the one in the last breakpoint, then it doesn't write out that media query.

yvonnetangsu replied:
`    /// Only write media query if $i = 1 or if $current_base is different from $previous_base.
    @if $current-base != $previous-base {
      $key: nth($bp-list, $i);`
Line 40 in the mixin 😃

sherakama replied:
Ah I see. Thanks for clarifying that.

sherakama commented on core/src/scss/utilities/variables/core/_grid.scss:
```diff
@@ -38,6 +38,18 @@ $su-grid-media: (
+/// @name $su-grid-device
+///
+/// @group variable
+$su-grid-device: (
```
handy.

sherakama commented on core/src/scss/utilities/functions/modular-scale/_get-modular-spacing-value.scss:
```diff
@@ -0,0 +1,21 @@
+@function get-modular-spacing-value($step: 0, $device: 'desktop') {
+  $value: map-deep-get($su-modular-spacing-map, $step, $device);
+  @if ($step < -4 and $step > 10) {
+    @warn 'Modular spacing step value must be an integer between -4 and 10.';
```
This could be: 
```suggestion
    @error 'Modular spacing step value must be an integer between -4 and 10.';
```

sherakama commented on core/src/scss/utilities/functions/modular-scale/_get-modular-spacing-value.scss:
```diff
@@ -0,0 +1,21 @@
+    @warn 'Modular spacing step value must be an integer between -4 and 10.';
+  }
+  @if $value == null {
+    @warn 'Could not find this spacing value in the $su-modular-spacing-map. Check the parameters.';
```
```suggestion
    @error 'Could not find this spacing value in the $su-modular-spacing-map. Check the parameters.';
```

sherakama commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr704-artqrbwckdrvxjsngf6uaw9xrixwdmzq.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5f78290f8ff35823c4ef06f8

sherakama commented:
The big question for this PR is what to do with the changing modular-spacing values. Ideally, I would like the modular-spacing function we have today to be updated with the new values. If that cannot be accomplished we need to discuss whether to create a new function and deprecate the existing one, move the new spacing to version 7, or try to split the difference by creating some sort of backwards compatibility within the function itself.

yvonnetangsu commented:
> The big question for this PR is what to do with the changing modular-spacing values. Ideally, I would like the modular-spacing function we have today to be updated with the new values. If that cannot be accomplished we need to discuss whether to create a new function and deprecate the existing one, move the new spacing to version 7, or try to split the difference by creating some sort of backwards compatibility within the function itself.

Yes, agree! Sounds like some group discussions would be needed. I might use this Decanter branch on a local site just to test it out and see the effect.

sherakama commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr704-artqrbwckdrvxjsngf6uaw9xrixwdmzq.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5f78290f8ff35823c4ef06f8

sherakama commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr704-artqrbwckdrvxjsngf6uaw9xrixwdmzq.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5f78290f8ff35823c4ef06f8

yvonnetangsu commented:
@sherakama @katrialesser Per group discussion today, I'm not replacing the current @modular-spacing mixin, instead created a new one @responsive-spacing. Since we're not replacing the old one, I'm eliminating step -3 and -4 (which we never used anyway and doesn't make much sense to me since the pixel values are basically the same for all breakpoints). No rush on this - just a FYI to take a look when you have a chance. Thanks!

sherakama commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr704-artqrbwckdrvxjsngf6uaw9xrixwdmzq.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5f78290f8ff35823c4ef06f8

yvonnetangsu commented:
@sherakama Kerri has confirmed the spacing values. You think we could merge this in and punch a release? I could use this for the Giving site 😄  
The cc issues are with the sass map spacing - I kept it consistent with our other sass maps.

sherakama commented:
@yvonnetangsu I've added the deprecated notices and changed `modular-scale` to `responsive-scale` everywhere it was used. Please have a review.

yvonnetangsu commented:
> @yvonnetangsu I've added the deprecated notices and changed `modular-scale` to `responsive-scale` everywhere it was used. Please have a review.

Awesome - thanks for the work. I'll review a bit later today.

sherakama commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr704-artqrbwckdrvxjsngf6uaw9xrixwdmzq.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5f78290f8ff35823c4ef06f8

yvonnetangsu commented on core/src/scss/utilities/mixins/modular-scale/_modular-spacing.scss:
```diff
@@ -13,15 +13,19 @@
 /// @param {string} $type  Spacing type, either 'padding' or 'margin'.
 /// @param {list} $steps - List (1 to 4 values): Either integer (from -4 to 9 including 0) or null. Follows the shorthand notation of padding/margin. Use null to skip a side. A value of 0 returns the modular spacing base value for that breakpoint.
 ///
+/// @deprecated modular-spacing mixin has been deprecated font will be removed in Decanter Version 7. Please replace with responsive-spacing instead.
```
```suggestion
/// @deprecated modular-spacing mixin has been deprecated and will be removed in Decanter Version 7. Please replace with responsive-spacing instead.
```

yvonnetangsu replied:
@sherakama Just this one typo. All your changes look great to me. I looked at the tugboat build and the components all look fine to me - nothing breaks AFAIK. Thank you for replacing the instances!

sherakama reviewed (approved):
GTG

yvonnetangsu commented:
> GTG

Thanks for your work and review! Yay 🥳
